book/src/non-system-libraries.md

```diff
@@ -0,0 +1,106 @@
+Now let's suppose we want to generate bindings for a non-system library. We
+will be the same crate setup as the previous tutorial. First let's create a new
+directory `hello` with two files inside it. A `c` source file `hello.c`
+containing
+```c
+int hello() {
+    return 42;
+}
+```
+and a `c` header file `hello.h` containing
+```c
+int hello();
+```
+given that the library has not been compiled yet, we need to modify the
+`build.rs` build script to compile the `hello.c` source file into a static
+libary:
+
+```rust
+extern crate bindgen;
+
+use std::env;
+use std::path::PathBuf;
+
+use bindgen::CargoCallbacks;
+
+fn main() {
+    // This is the directory where the `c` library is located.
+    let libdir_path = PathBuf::from("hello")
+        // Canonicalize the path as `rustc-link-search` requires an absolute
+        // path.
+        .canonicalize()
+        .expect("cannot canonicalize path");
+
+    // This is the path to the `c` headers file.
+    let headers_path = libdir_path.join("hello.h");
+    let headers_path_str = headers_path.to_str().expect("Path is not a valid string");
+
+    // This is the path to the intermediate object file for our library.
+    let obj_path = libdir_path.join("hello.o");
+    // This is the path to the static library file.
+    let lib_path = libdir_path.join("libhello.a");
+
+    // Tell cargo to look for shared libraries in the specified directory
+    println!("cargo:rustc-link-search={}", libdir_path.to_str().unwrap());
+
+    // Tell cargo to tell rustc to link our `hello` library. Cargo will
+    automatically know it must look for a `libhello.a` file.
+    println!("cargo:rustc-link-lib=hello");
+
+    // Tell cargo to invalidate the built crate whenever the header changes.
+    println!("cargo:rerun-if-changed={}", headers_path_str);
+
+    // Run `clang` to compile the `hello.c` file into a `hello.o` object file.
+    // Unwrap if it is not possible to spawn the process.
+    if !std::process::Command::new("clang")
```

Comment:
Do we want to recommend just using `clang`? I think we probably want to point to the `cc` crate or so, which should deal with a lot of edge cases.

Reply:
@emilio I think using the `cc` crate is not very helpful because it handles all the linking for you, which means you don't need to specify anything using `cargo:rustc-link-*`. At the same time anyone compiling a more complex project couldn't rely on it and would have to do some custom invocation like we did here iiuc.